specs/yoga/clusterapi-driver.rst

```diff
@@ -0,0 +1,200 @@
+|                          |                      | code. Defaults to rolling |
+|                          |                      | upgrade (like Magnum).    |
++--------------------------+----------------------+---------------------------+
+| Self-healing             | Partial / uncertain. | Supported with            |
+|                          |                      | infrastrcuture agnostic   |
+|                          |                      | code via reconciliation   |
+|                          |                      | loop.                     |
++--------------------------+----------------------+---------------------------+
+| Auto-scaling             | Supported for        | Supported with            |
+|                          | default node group.  | infrastructure agnositc   |
+|                          |                      | code.                     |
++--------------------------+----------------------+---------------------------+
+| Multiple node groups     | Supported.           | Supported, with no default|
+|                          |                      | group.                    |
++--------------------------+----------------------+---------------------------+
+| Additional networks      | Supported (review    | Supported                 |
+|                          | pending)             |                           |
++--------------------------+----------------------+---------------------------+
+| New Kubernetes versions  | Test burden on Magnum| Test burden split between |
+|                          | entirely.            | Cluster API for           |
+|                          |                      | Kubernetes, Magnum for    |
+|                          |                      | infrastructure provision. |
++--------------------------+----------------------+---------------------------+
+
+The intention for the new driver is to make it compatible with the
+Magnum REST API.
+
+Alternatives
+------------
+
+A roadmap securing ongoing Magnum development is important for the
+project.  Alternatives would include doing something similar outside
+of Magnum.
+
+Implementation
+==============
+
+Cluster API is not a drop-in replacement for a Magnum COE driver, and some
+further work will be needed to ensure it has the resources it requires in
+order to function.  Cluster API provisions and operates Kubernetes clusters
+from a dedicated management cluster, which requires access to OpenStack
+APIs and provisioned infrastructure.
+
+Deployment of the management cluster is a significant aspect of
+driver implementation.  This could be performed in two ways:
+
+* Boostrap a Kubernetes cluster in a service tenancy on infrastructure
+  hosted by OpenStack.
+* Deploy Cluster API's management services to an existing "provider
+  cluster".
+
+Each approach could be valid in different configurations.  The
+bootstrap cluster could feasibly be implemented outside of Magnum's
+scope by OpenStack deployment frameworks.
```

Comment:
Can we be stronger than this...

Bootstrapping the management cluster is out of scope, the driver simply depends on Cluster API being installed in a similar way to expecting Heat API to be installed.

However, we expect to integrate into devstack a method that uses k3s to provide a simple management cluster that can be used to test this driver. A k3s based Cluster API can its-self be used to bootstrap an HA management cluster.

I guess the key thing we need to state here is that we expect to add a driver alongside the existing drivers, and it will only support kubernetes. I think I went for an os_type of "cluster api", and a coe_type of "kubernetes" in my current patch I put upstream.